Firestore/core/src/firebase/firestore/core/query.h

```diff
@@ -36,28 +42,46 @@ class Query {
    * @return A new instance of Query.
    */
   static Query AtPath(model::ResourcePath path) {
-    return Query(std::move(path));
+    return Query(std::move(path), std::vector<std::unique_ptr<core::Filter>>());
   }
 
   /** Initializes a query with all of its components directly. */
-  explicit Query(model::ResourcePath path /* TODO(rsgowman): other params */)
-      : path_(std::move(path)) {
-  }
+  Query(model::ResourcePath path,
```

Comment:
I'm increasingly convinced that this approach to immutability is needlessly costly.

If not for this incremental build up, Query could easily contain a `std::vector<std::unique_ptr<Filter>>`. Instead we're `std::shared_ptr`ing to at least avoid making this comically bad at the cost of this bleeding out elsewhere, as in my recommendation that `Filter::Create()` just return `std::shared_ptr` directly.

If we had a `Query::Builder` that built up state on a mutable vector of filters and then `std::move()`d that into an immutable `Query` instance on `Build()` we'd get the same effect without all the waste.

Alternatively, we could just do as "normal" C++ does and make getters const and mutators non-const and just flip to const instances and references once a `Query` is complete.

WDYT? Should we put this on this week's API discussion?

Reply:
We decided against this.